### PR TITLE
Generate predictable UUIDs for offline accounts

### PR DESCRIPTION
**Pull request by Scrumplex**

### CODE REVIEW

**Comment by TheKodeToad** [comment]
Won't apply to Launch Offline though. can you fix that?

**Comment by Scrumplex** [comment]
Launch Offline should use the UUID attached to the active account

**Comment by TheKodeToad** [comment]
...Which probably isn't what most people want, right?

**Comment by Scrumplex** [comment]
> ...Which probably isn't what most people want, right?

Uh they do? I would want to keep my inventory when I join a single player world with the same account, regardless if it was online or offline.

**Comment by TheKodeToad** [comment]
You keep your inventory regardless of UUID on singleplayer. If you e.g. want to test on your own server using offline mode, you might not be able to with a single account (not tested).

**Comment by Scrumplex** [comment]
The whole point of this PR is to make UUIDs of offline accounts predictable. I would argue that Play Offline is already very predictable, as is.

**Comment by TheKodeToad** [comment]
> The whole point of this PR is to make UUIDs of offline accounts predictable. I would argue that Play Offline is already very predictable, as is.

Not really, since you can enter the same username and get a different uuid

File: launcher/minecraft/auth/MinecraftAccount.cpp [hunk 1]
```diff
@@ -334,3 +335,32 @@ void MinecraftAccount::incrementUses()
...
+        array[index] |= value;
+    };
+    auto bAnd = [](QByteArray& array, qsizetype index, char value) {
+        array[index] &= value;
+    };
+#endif
+    bAnd(digest, 6, (char) 0x0f); // clear version
+    bOr(digest, 6, (char) 0x30); // set to version 3
+    bAnd(digest, 8, (char) 0x3f); // clear variant
+    bOr(digest, 8, (char) 0x80); // set to IETF variant
```
Comment (TheKodeToad) [inline launcher/minecraft/auth/MinecraftAccount.cpp:363]:
```suggestion
    digest[6] &= (char) 0x0f; // clear version
    digest[6] |= (char) 0x30; // set to version 3
    digest[8] &= (char) 0x3f; // clear variant
    digest[8] |= (char) 0x80; // set to IETF variant
```

Why doesn't this work? it's more readable

**Comment by TheKodeToad** [comment]
anyway, the current system can be controlled by the user more :woman_shrugging:
(I also don't want a dreaded github argument)

File: launcher/minecraft/auth/MinecraftAccount.cpp (hunk 1 above)
Comment (Scrumplex) [reply launcher/minecraft/auth/MinecraftAccount.cpp:363]: Doesn't work because Qt 5 returns a QCharRef that doesn't support the `|` and `&` operators.

**Comment by Scrumplex** [comment]
> Not really, since you can enter the same username and get a different uuid

nope. This uses the md5 to generate a hash and just represents that as a UUID

This is completely deterministic.

**Comment by TheKodeToad** [comment]
> > Not really, since you can enter the same username and get a different uuid
> 
> nope. This uses the md5 to generate a hash and just represents that as a UUID
> 
> This is completely deterministic.

I mean with "Launch Offline"

**Comment by Scrumplex** [comment]
If you launch an instance using `Launch Offline` Prism will just take the UUID that we locally cached in `accounts.json`. See https://github.com/PrismLauncher/PrismLauncher/blob/01afa8957f477d347c5103305a038a30edd24ca2/launcher/minecraft/auth/MinecraftAccount.cpp#L302

So if I click Launch Offline with my usual MSA account, I will have the UUID `f2873756-429d-413a-b22d-6a976ed0d3f2` no matter the username I chose. This is intended behavior.

UUIDs ingame:

Launch (Online):
```
/data get entity Scrumplex UUID
Scrumplex has the following entity data: [I; -226019498, 1117602106, -1305646441, 1859179506]
```

Launch Offline (same username):
```
/data entity Scrumplex UUID
Scrumplex has the following entity data: [I; -226019498, 1117602106, -1305646441, 1859179506]
```

Launch Offline (different username):
```
/data get entity Secondplex UUID
Secondplex has the following entity data: [I; -226019498, 1117602106, -1305646441, 1859179506]
```

**Comment by TheKodeToad** [comment]
What I mean is if you change the selected account, the uuid will change when using Launch Offline.

File: launcher/minecraft/auth/MinecraftAccount.cpp (hunk 1 above)
Comment (TheKodeToad) [reply launcher/minecraft/auth/MinecraftAccount.cpp:363]: oops

**Review by TheKodeToad — APPROVED** [review]

**Comment by TayouVR** [comment]
looks good, though won't approve yet because I haven't tested it

**Comment by TheKodeToad** [comment]
I tested, and it had the same results for `Player579` (yes, I somehow remembered that)

**Review by TayouVR — APPROVED** [review]

**Pull request merged**